CRM/Core/Form/EntityTrackingTrait.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * This trait provides a form of lazy loading for forms.
+ *
+ * It is intended to reduce the need for forms to pass values around while avoiding
+ * constantly reloading them from the database.
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC https://civicrm.org/licensing
+ */
+trait CRM_Core_Form_EntityTrackingTrait {
+
+  /**
+   * Array of loaded entities.
+   *
+   * These are stored with keys that are consistent with apiv4 style parameters.
+   *
+   * @var array
+   */
+  private $entities = [];
+
+  public function setEntity($entity, $identifier, $values): void {
+    $this->entities[$entity][$identifier] = $values;
+  }
+
+  /**
+   * Get the value for a property of an entity, loading from the database if needed.
+   *
+   * Permissions are not applied to the api call.
+   *
+   * @param string $entity
+   * @param int $id
+   * @param string $key
+   *
+   * @api supported for use outside of core. Will not change in a point release.
+   *
+   * @return mixed
+   */
+  public function getEntityValue(string $entity, int $id, string $key) {
+    if (!isset($this->entities[$entity][$id]) || !array_key_exists($key, $this->entities[$entity][$id])) {
```

Comment:
so I suppose $identifier in L34 may not be an actual int but here we are assuming it will be an int right?

Reply:
hmm yeah - I think I go back & forth on whether we should only support ids or whether we should support string indentifiers too...

Reply:
ok - all ints now

Reply:
Actually I think it's quite useful to be able to look up stuff by name as well as id.